fix: revise vscode extension directories

xai opened the pull request:
<!--
Thank you for your Pull Request. Please provide a description and review
the requirements below.

Contributors guide: https://github.com/theia-ide/theia/blob/master/CONTRIBUTING.md

Note: Security vulnerabilities should not be disclosed on GitHub, through a PR or any
other means. See SECURITY.md at the root of this repository, to learn how to report
vulnerabilities.
-->

#### What it does

<!-- Include relevant issues and describe how they are addressed. -->
This patch eliminates the use of the temporary directories `vscode-unpacked` and `vscode-copied` for installing and deploying vscode extensions. The file-handler and directory handler for vscode extensions have been adjusted accordingly:

1. A temporary directory is now only used for downloading extensions from the registry. This temporary directory is now user-specific and resides within the configdir (i.e., per default in the user's home: <userhome>/.theia/tmp/) to avoid clashes and permission issues on multi-user operating systems that share temporary directories, such as Linux or BSDs. Having this temporary directory in a location that is configurable by the user also seems the more sensible approach when extensions are considered confidential data.
2. `$configDir/deployedPlugins` replaces our volatile `/tmp/vscode-copied` deployment directory. Having a more permanent way of handling installed extensions should improve startup time and reduce issues with multiple instances running in parallel.
3. The local file resolver unpacks the vsix file from the temp dir into `$configDir/deployedPlugins/<extension-id>`.
4. The simplified directory handler loads unpacked extensions directly from `$configDir/deployedPlugins/<extension-id>`.
5. We use `$configDir/extensions` as a location for the user to drop vsix files that will be installed to the deployment location automatically on startup. We do not manage or remove files within `$configDir/extensions`.

Overall, this should improve the stability on systems with shared temp dir locations and reduce the startup of the first application start after a reboot.

Fixes #12757

Contributed on behalf of STMicroelectronics

#### How to test

<!-- Explain how a reviewer can reproduce a bug, test new functionality or verify performance improvements. -->

Installing from registry

1. Start Theia
2. Use the vsix registry to install an extension
3. Verify that the extension is loaded and its functionality is enabled.
4. Verify that the extension is permanently stored unpacked in $configDir/deployedPlugins/<extension-id>.
5. Try starting Theia again to verify the extension gets correctly loaded
6. Remove the extension (this should indicate that a reload of the window is required), and verify that the extension is no longer active
7. Verify that $configDir/extensions/<extension-id> has been removed

Installing a local vsix extensions

1. Start Theia
2. Select `Install from VSIX` in the triple-dot menu of the extensions view
3. Browse to a local .vsix file and install ut
4. Verify that the extension is permanently stored unpacked in $configDir/deployedPlugins/<extension-id>.
5. Remove the extension (this should indicate that a reload of the window is required), and verify that the extension is no longer active
6. Verify that $configDir/extensions/<extension-id> has been removed

Store a vsix file in the drop-in location

1. Put a .vsix file into $configDir/extensions
2. Start Theia
3. Verify that the extension is permanently stored unpacked in $configDir/deployedPlugins/<extension-id>.
4. Removing the extension will cause it to be reinstalled until you remove it from the drop-in location

#### Follow-ups

<!-- Please list potential follow-up work, including known issues, possible future work, identified technical debt, and potentially introduced technical debt. If the PR introduces technical debt, specify the reason why this is acceptable. Please create tickets and link them here. Please use the label "technical debt" for new issues when it applies. -->

#### Review checklist

- [x] As an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

tsmaeder commented:
I have a couple of general comments:

1. We should not download extensions to a shared directory. I have heard of use cases where even the existence of certain extensions may be a trade secret. Isn't there a user-specific tmp on linux?
2. We should never move or delete files that are managed by the user. So if the user puts a *.vsix into `.theia/extensions`, we should not move nor delete it from there.
3. If we download *.vsix files as a side-effect of an installation, we should clean up those files once we no longer need them. Relying on the OS or user to clean up files created by software is bad practice, IMO.
4. I think it would be a good idea to separate "drop in locations" and "deployment locations". Any given directory should be one or the other. If an application does not have the ability to install additional plugins, we should never deal with drop-in locations. Concretely, this would mean that we have `.theia/extensions` as a drop-in location where the user can put extensions and `.theia/deployedPlugins` where stuff is extracted to.

xai commented:
Thanks, @tsmaeder, for the quick response!

>     1. We should not download extensions to a shared directory. I have heard of use cases where even the existence of certain extensions may be a trade secret. Isn't there a user-specific tmp on linux?

I don't think there is a consistent user-specific temp dir in Linux. According to `man 3 tmpfile`:
``` 
The  standard  does  not  specify  the  directory  that  tmpfile()  will  use.  Glibc will try the path prefix P_tmpdir defined in <stdio.h>, and if that fails the directory /tmp.
```
The shared /tmp is often also on a different filesystem/mountpoint than the user home. In several scenarios, the user home will be better secured (e.g., encrypted with user-specific credentials, mounted after kerberos auth, ...) than the shared temp dir location.

In our use case, the download directory is meant to be user-specific (that's why I encoded the username into it), but sure, the default permissions will most likely be world-readable (755) if we do not change them. So I see two options:
1. Make sure the permission of /tmp/vscode-download-$username is 700
2. Download to a path within the user home, such as .theia/tmp/download

I think if we want to treat extensions as potential secrets, then the second approach is more sensible, considering the user home to be a much safer place than /tmp. Of course, this not being a temp dir that the OS handles, we have to clean it up ourselves.
So what do you think about `<userhome>/.theia/tmp/download-<randomuuid/processid>`?

>     2. We should never move or delete files that are managed by the user. So if the user puts a *.vsix into `.theia/extensions`, we should not move nor delete it from there.

ACK.

>     3. If we download *.vsix files as a side-effect of an installation, we should clean up those files once we no longer need them. Relying on the OS or user to clean up files created by software is bad practice, IMO.

That makes sense, especially if we decide to use a location within the user's home. We have to be careful here with regard to robustness in case of multiple Theia processes (run by the same user). So, cleaning up after successful unpacking is easy enough. What do we do if the application gets killed hard by the user or the OOM killer and our cleanup is not finished? Should we try to clean at the application start? Then we have to make sure not to remove stuff that another running Theia process is currently using.

>     4. I think it would be a good idea to separate "drop in locations" and "deployment locations". Any given directory should be one or the other. If an application does not have the ability to install additional plugins, we should never deal with drop-in locations. Concretely, this would mean that we have `.theia/extensions` as a drop-in location where the user can put extensions and `.theia/deployedPlugins` where stuff is extracted to.

Makes sense; this is backward-compatible and avoids unwanted interactions between our unpacking/deployment and the user. I'll revise the code accordingly.

xai commented:
I revised my PR to use

* `~/.theia/deployedPlugins/` as a permanent location for installed extensions. This replaces the volatile `/tmp/vscode-copied`
* `~/.theia/extensions/` as a drop-in location that the user manages (we do not touch anything there) and that only serves as a source of extensions that will be installed on startup
* `~/.theia/tmp/` as a replacement for the remaining shared temp directories. It can be used to download .vsix files from the registry (they do not belong in ~/.theia/extensions`, as we no longer manage this directory).

tsmaeder commented on packages/plugin-ext-vscode/src/node/plugin-vscode-decompress.ts:
```diff
@@ -0,0 +1,36 @@
+    try {
+        await decompress(sourcePath, destPath);
+        if (sourcePath.endsWith('.tgz')) {
+            const extensionPath = path.join(destPath, 'package');
```
A comment to explain this path mangling would be nice

xai replied:
I just moved this part of code over there from https://github.com/eclipse-theia/theia/blob/2d30b295c1158f401d63591eb67b46e93cd6f318/packages/plugin-ext-vscode/src/node/plugin-vscode-file-handler.ts#L76

I had no idea for which exact use case we needed this, but I did not want to create regressions, so I kept it. Comes from [this commit](https://github.com/eclipse-theia/theia/commit/ccaa5d84c5718ce66e887b7bed620366fc859c11) originally. It seems that it unpacks an additional `vscode_node_modules.zip` file in case the source file was a .tgz. I can add a comment that says so, but I do not know the exact use case.

tsmaeder commented on packages/plugin-ext-vscode/src/node/plugin-vscode-deployer-participant.ts:
```diff
@@ -25,8 +27,21 @@ export class PluginVSCodeDeployerParticipant implements PluginDeployerParticipan
-        const extensionsDirUri = await this.environments.getExtensionsDirUri();
-        context.userEntries.push(extensionsDirUri.withScheme('local-dir').toString());
+        const extensionDeploymentDirUri = await this.environments.getDeploymentDirUri();
+        context.userEntries.push(extensionDeploymentDirUri.withScheme('local-dir').toString());
```
Is this releated to the `--plugins=local-dir:../../plugins` cli parameter? Since we separated "drop-in" directories from "deployment" directories, what does the cli-parameter now mean?

xai replied:
I don't think this part of the code ever cared about `--plugins`:
https://github.com/eclipse-theia/theia/blob/2d30b295c1158f401d63591eb67b46e93cd6f318/packages/plugin-ext-vscode/src/common/plugin-vscode-environment.ts#L31
Maybe this is a difference between theia plugins and vscode extensions? For vscode extensions, it looks already hard-coded to me.

tsmaeder commented on packages/plugin-ext-vscode/src/node/plugin-vscode-decompress.ts:
```diff
@@ -0,0 +1,36 @@
+        }
+        return true;
+    } catch (error) {
+        console.error(`Failed to decompress ${sourcePath} to ${destPath}: ${error}`);
```
Why are we not just letting the exception propagate?

xai replied:
Sure, stupid mistake.

tsmaeder commented on packages/plugin-ext-vscode/src/node/plugin-vscode-file-handler.ts:
```diff
@@ -50,34 +39,45 @@ export class PluginVsCodeFileHandler implements PluginDeployerFileHandler {
+
+            console.log(`[${id}]: decompressed`);
+        } catch (error) {
+            console.error(`[${id}]: error while decompressing`, error);
```
Why not let the error propagate? How is the invoker going to tell sometihing failed?

tsmaeder commented on packages/plugin-ext-vscode/src/node/plugin-vscode-file-handler.ts:
```diff
@@ -50,34 +39,45 @@ export class PluginVsCodeFileHandler implements PluginDeployerFileHandler {
+            const extensionDeploymentDirPath = FileUri.fsPath(deployedPluginsDirUri.resolve(pluginId));
+            return extensionDeploymentDirPath;
+        } catch (error) {
+            console.error('Error resolving extension directory:', error);
```
If the error is unhandled, we'll get a stack trace. If it isn't shouldn't the handler print the error?

tsmaeder commented on packages/vsx-registry/src/node/vsx-extension-resolver.ts:
```diff
@@ -74,16 +75,24 @@ export class VSXExtensionResolver implements PluginDeployerResolver {
-        console.log(`[${resolvedId}]: trying to download from "${downloadUrl}"...`, 'to path', downloadPath);
-        if (!await this.download(downloadUrl, extensionPath)) {
+        const downloadDir = await this.getTempDir();
+        await fs.ensureDir(downloadDir);
```
Do we clean this up somewhere?

xai replied:
Now we do ;)
I adjusted the `PluginVsCodeFileHandler` to cleanup `pluginEntry().originalPath()` if it is a temp dir.

tsmaeder reviewed (changes requested):
Mostly comments, but I think we should fix the "move" logic.

tsmaeder commented:
Some notes from testing:

1. Dropping the "vscode-drawio" plugin into the "extensions" folder inside the `~/.theia` folder blocks the UI at startup for 20 seconds with a white screen. When we install extensions via the UI, the extensions can startup after being installed, so there's no reason to block while we wait for the plugins to install. 
This happens on **every** startup, not just the first. 
3. After startup, we end up with an "uninstall" button for the extesnsion in the plugins view. When you click it, the extension is uninstalled. However, the extension is reinstalled the next time we start Theia. Maybe we should consider extensions installed via "drop-in" folder as "built-in" and the way to remove them would be to remove the *.vsix from the drop-in folder?

I can kind of live with the second behavior, but Nr. 1 is a must-fix, IMO.

xai commented on packages/plugin-ext-vscode/src/node/local-vsix-file-plugin-deployer-resolver.ts:
```diff
@@ -38,28 +40,70 @@ export class LocalVSIXFilePluginDeployerResolver extends LocalPluginDeployerReso
+    protected async getExtensionDeploymentDir(pluginId: string): Promise<string> {
+        try {
+            const deployedPluginsDirUri = await this.environment.getDeploymentDirUri();
+            const normalizedExtensionId = filenamify(pluginId, { replacement: '_' });
```
"Should not happen" ™, as the `pluginId` that is passed in there is parsed from the package.json after unpacking. This is btw. the reason why the application is starting slow when drop-in vsix files are used currently. We always unpack to parse the plugin-id from the package.json. This is a problem ...

xai commented on packages/plugin-ext-vscode/src/common/plugin-vscode-environment.ts:
```diff
@@ -24,13 +24,32 @@ export class PluginVSCodeEnvironment {
+        if (!this._tmpDirUri) {
+            const configDir: URI = new URI(await this.environments.getConfigDirUri());
+            const processId: string = process.pid.toString();
+            this._tmpDirUri = configDir.resolve('tmp').resolve(`${prefix ?? 'tmp-'}-${processId}`);
```
If I understood this correctly, this will use the operating system's temp dir mechanism, that we considered unsafer than a location within the user's conf dir.

xai replied:
I'll try.

xai commented:
>     1. Dropping the "vscode-drawio" plugin into the "extensions" folder inside the `~/.theia` folder blocks the UI at startup for 20 seconds with a white screen. When we install extensions via the UI, the extensions can startup after being installed, so there's no reason to block while we wait for the plugins to install.
>        This happens on **every** startup, not just the first.

It should be fixed now. I reverted back to using the filename of a dropped vsix file as extension-id instead of *always* unpacking and parsing package.json. Would be nicer to have the correct extension id, but we can deal with this in a future PR, I'd be happy to look into this.


>     2. After startup, we end up with an "uninstall" button for the extesnsion in the plugins view. When you click it, the extension is uninstalled. However, the extension is reinstalled the next time we start Theia. Maybe we should consider extensions installed via "drop-in" folder as "built-in" and the way to remove them would be to remove the *.vsix from the drop-in folder?

Treating them as "built-in" would be in line with the user being the only one who manages them. This would probably also fit well with having a lookup mechanism for extensions because then we could track the source of deployed extensions (e.g., downloaded, dropped-in, installed-from-file) and treat them accordingly. 

> I can kind of live with the second behavior, but Nr. 1 is a must-fix, IMO.

Nr. 1 should be done now. Nr. 2 is definitely a good candidate for a follow-up ticket.

xai commented on packages/plugin-ext-vscode/src/node/local-vsix-file-plugin-deployer-resolver.ts:
```diff
@@ -38,28 +40,70 @@ export class LocalVSIXFilePluginDeployerResolver extends LocalPluginDeployerReso
+        await fs.move(tempUnpackDir, extensionDeploymentDir, { overwrite: true });
+
+        // if temp dir is empty, remove it
+        if (fs.readdirSync(tempDir).length === 0) {
```
Yes, that was my thought here. I removed all Sync operations now and hope that the error handling is sufficient.

tsmaeder reviewed (commented):
Code looks good to me now, but I still get a 20s pause with a blank window the first time I drop the draw.io vsix into the `.theia/extensions` folder.

tsmaeder commented:
@xai we have a conflict.

xai commented:
> @xai we have a conflict.

Resolved.
